### ISSUE TITLE
url encode stateparams

### ISSUE THREAD

**Issue by pdeva** [angular-ui/ui-router#1119]
If i pass a string like `/default` to `$stateParams`, it seems to mess up the routing:

    var name = "/default";
    this.$state.go("app.transactions.detail", { txnName: name });

While the controller is invoked, the url is unable to switch to.
when i encode the value before passing it to $stateparams, it works.

however, ui-router should take care of doing the encoding/decoding for me.

**Comment by elmalto** [comment]
@panuhorsmalahti great, thank you very much!

**Comment by martin-brennan-csnet** [comment]
Is this supposed to work with queryString parameters as well? I've tried the custom types with `$urlMatcherFactory` and it works with parameters in the url e.g. `/route/{param:MyType}/etc` but not with queryString parameters e.g. `/route?{param:MyType}`.

**Comment by christopherthielen** [comment]
@martin-brennan-csnet yes, it is expected to work for query parameters too.

**Comment by soarb** [comment]
@christopherthielen I appear to have a similar issue to @martin-brennan-csnet. Registering a custom type for my /route/{url:nonURIEncoded} parameter works a treat, but as a queryString parameter even though a call is made to my custom types 'encode' method, the forward slashes are still being converted to %2F when I call $state.go('state', {url: 'http://someurl/somepath'}) - am I missing something else here? Cheers! Ben

**Comment by soarb** [comment]
@christopherthielen Oh I see. It would appear that 'encodeURIComponent' is still invoked on whatever my custom type's 'encode' method returns, therefore the forward slashes are still encoded. Is there any way to disable this? Thanks!

**Comment by mismith** [comment]
I still can't get this to work. Am I missing something? Here's a quick plunk showing the problem:
http://plnkr.co/edit/tNlRmadZRUGRPHhrKqTQ?p=preview

**Comment by mismith** [comment]
It looks like the problem lies in the `UrlMatcher.format()` method, where `encodeURIComponent()` gets called on the result of the custom `urlMatcherFactory.type`s `encode()` method no matter what. I'm not sure I'm seeing the big picture here as to why this occurs; maybe @nateabele could help clarify? Is this the intended behaviour? Shouldn't the URI encoding be left to the customizable `encode`?

**Comment by anpsince83** [comment]
I'm seeing the same behavior as @mismith with ui-sref. Is there any way to have the ui-sref not encode a specific parameter?

**Comment by andrewburgess** [comment]
Just to expand on @mismith 's comment above:
It looks like the encoding happens here: https://github.com/angular-ui/ui-router/blame/985c57a5a6f49edb520f56b6cd7f4a980839d19e/src/urlMatcherFactory.js#L347

even though the parameter is supposedly encoded here: https://github.com/angular-ui/ui-router/blame/985c57a5a6f49edb520f56b6cd7f4a980839d19e/src/urlMatcherFactory.js#L338

I agree that the `param.type.encode` method should handle the `encodeURIComponent` call itself, and not be universally applied as it currently is.

**Comment by fabiosussetto** [comment]
@anpsince83 For my use case, I have to support dynamically generated urls for a set of pages (coming from a CMS), this is how I solved it:

```javascript
.config(function ($stateProvider, $urlRouterProvider, $urlMatcherFactoryProvider) {
  function valToString(val) {
    return val !== null ? val.toString() : val;
  }

        $urlMatcherFactoryProvider.type('nonURIEncoded', {
            encode: valToString,
            decode: valToString,
            is: function () { return true; }
        });

  $stateProvider
          .state('main.sitemapPage', {
                    url: '{url:nonURIEncoded}',
                    templateUrl: 'sitemap/page.html',
                    controller: 'SitemapPageCtrl',
                    controllerAs: 'pageCtrl'
                });

```

I added this dynamic state as the last one, so that before I can match other static ones. This state supports urls such as /my-section/subpage/page.

Hope it's helpful.

**Comment by anpsince83** [comment]
@fabiosussetto Thanks, I'm doing something similar to that as well. That allows the unencoded '/' in the url displayed in the browser url bar, which is great. However, the issue I'm still seeing, that @mismith noted, is that even with a custom type defined that handles the encoding/decoding, the ui-sref directive still uirencodes the paramater value, so you end up with an href on a link that looks like this: href='/my%2fparam'. Which is undesirable. It would be nice if as @andrewburgess mentioned, the `param.type.encode` method could handle the encoding.

**Comment by jylinman** [comment]
+1 for @anpsince83 's reply.  When I render something like `<a ui-sref="view({ slug: item.slug })">` I still get the encoding.  Any thoughts on how to bypass this?

**Comment by JohannesFerner** [comment]
@jylinman or @fabiosussetto could you provide a working example on fiddle or codepen?

**Comment by jylinman** [comment]
Yeah sure, here you go.

http://jsfiddle.net/jylinman/rm1Lbptf/1/

**Comment by JohannesFerner** [comment]
@jylinman thx. but i think we have a different requirement here: 
i want to have something like ui-sref="state({path: 'a/b/c/d/f/'});
resulting in 
/domain/paht/a/b/c/d/f/

in your fiddle i would get a-f url-encoded like: '#/1%2F2%2F3%2F4!' instead of '#1/2/3/4/5/'

i resolved my case to work with those urls by using url: 'test/{test:.*}' but ui-sref is still encoding it (with and without using an own urlmatcherFactory Type)

**Comment by jylinman** [comment]
Yes, that's what my fiddle was doing-- I have two examples, one set by a `$state` and the other by `ui-sref`.  Just wanted to show where it shows up in my use cases.

**Comment by bora89** [comment]
@jylinman did you manage that encoding issue? Also wondering how get unencoded string from ui-sref

**Comment by JohannesFerner** [comment]
@andrewburgess @jylinman @bora89 i think we all have the same problem here: 
We want to have a `ui-sref="file({slug: 'A/B/C/D/E/F/});"`  to result in `<a href="file/A/B/C/D/E/F`instead of the url-encoded version. 

Has anyone of you had any progress on this issue?

**Comment by Vnthf** [comment]
when I called 
```
$state.go('main.common', {query : '/'}) 
```
It appears %252F
Is it fixed bug, right ? It doesn't work yet;;;;  
I think most of above comments can't resolve this issue except for custom code.

**Comment by ceoworks** [comment]
Hello, I'm having a problem with next url, which redirects me to default url, as in case "route not found": 
http://localhost:8000/#/user/restore/FW8AI4OLkNA+H/1WQLqzPA9VKgdzdCMDLASV81ZZRL0=

While my ui-router config is set to: 
```javascript
      .state('user.restore', {
        url: '/restore/:refreshToken',
        templateUrl: './app/components/forgot/new_password.html',
        controller: function (Network, $localStorage, authData) {
          var vm = this;
          console.log(authData);
          $localStorage.accessToken = authData.accessToken;
          $localStorage.refreshToken = authData.refreshToken;
          vm.accessToken = authData.accessToken;
          vm.refreshToken = authData.refreshToken;
        },
        controllerAs: 'restore',
        resolve: {
          authData: function (Network, $stateParams, $localStorage) {
            console.log('Storing RT to storage', $stateParams.refreshToken);
            $localStorage.refreshToken = $stateParams.refreshToken;
            return Network.checkRefreshToken();
          }
        }
      })
```

Also I was tried to set parameter as follows, but it didn't help:
```javascript
/{refreshToken}
/{refreshToken:[^/]*}
```
Router also continue to redirect, and doesn't get into "resolve: ..."

**Comment by circuitrider** [comment]
Same issue as ceoworks. This feature is related to performing password resets where a url encoded string is a token generated on the server received by the user in an email. User clicks link from their inbox and the routing fails and it goes to the default route if the token has an encoded forward slash in it.

**Comment by zivc** [comment]
People are struggling to grasp the simple concept of reading the issues before randomly copying and pasting code. This code works as of 25th August 2015, AngularJS v1.4.3, UI-Router version 0.2.15

```
<!DOCTYPE html>
<html ng-app="demo">

  <head>
    <script data-require="angular.js@*" data-semver="1.4.3" src="https://code.angularjs.org/1.4.3/angular.js"></script>
    <script data-require="ui-router@*" data-semver="0.2.15" src="//rawgit.com/angular-ui/ui-router/0.2.15/release/angular-ui-router.js"></script>
    <link rel="stylesheet" href="style.css" />
    <script src="script.js"></script>
  </head>

  <body>
    <div ng-controller="demoCtrl">
        {{path}}
        <div ui-view></div>
        <a ui-sref="system.folder({folder:'hello/lol/123/42/34234/234/sdsa'})">Click me</a>
    </div>
    <script>
        
      var demo = angular.module('demo', ['ui.router']);
      demo.config(['$stateProvider', '$urlRouterProvider', '$locationProvider', '$httpProvider', '$urlMatcherFactoryProvider',
          function($stateProvider, $urlRouterProvider, $locationProvider, $httpProvider, $urlMatcherFactoryProvider) {
            
            function valToString(val) { return val != null ? val.toString() : val; }
              function valFromString(val) { return val != null ? val.toString() : val; }
              function regexpMatches(val) { /*jshint validthis:true */ return this.pattern.test(val); }
              $urlMatcherFactoryProvider.type("MyType", {
                  encode: valToString,
                  decode: valFromString,
                  is: regexpMatches,
                  pattern: /[a-zA-Z0-9/\%]/
                });
                
            
            $stateProvider.state('system', {
                title:'System',
                url:'/',
                template:'<h1>System</h1> <div ui-view></div>'
            })
                .state('system.folder', {
                    title:'Folder',
                    url:'/{folder:MyType}',
                    template:'Folder: {{folder}}',
                    controller:['$scope',function($scope) {
                        console.log($scope);
                    }]
                });
        
            $urlRouterProvider.otherwise('/');
            
            
          }
      ]);
      demo.controller('demoCtrl', ['$scope','$location','$rootScope', function($scope,$location,$rootScope) {
        
        $rootScope.$on('$locationChangeStart', function() {
            $scope.path = window.location.href;  
        });
        
      }]);
    </script>
  </body>

</html>
```

Plnkr: http://run.plnkr.co/plunks/yrYxYUSs9xBqXzg6ZeM6/#/

**Comment by mismith** [comment]
Unfortunately, that doesn't fix the core issue @zivc 

While ui-router may navigate to the correct URL internally, the `/`s in the URL are still being URL encoded. See your code transposed into this simpler plunk:
http://plnkr.co/edit/CrD72FNYzLV1dATCjDdL?p=preview

This becomes an issue when, say, a user tries to open a `ui-sref`-set URL in a new tab, where those encoded slashes break `html5mode(true)` routing.

**Comment by zivc** [comment]
Ok, there is progress,
http://run.plnkr.co/plunks/srstfPWJ4ukhCMKLqsg4

In the address bar it shows the slashes OK, and on the page, but not in the status bar.

**Comment by mismith** [comment]
Sure, but unfortunately that still doesn't solve the problem. Try pumping out the result of that `$state.href()` call, [like I did in your plunker here](http://plnkr.co/edit/Air57UCelsGtq1Wal0Ia?p=preview), and you'll see that the slashes are still being URL encoded in Chrome and Firefox alike.

[As I mentioned earlier in this thread](https://github.com/angular-ui/ui-router/issues/1119#issuecomment-91007407), this occurs because of the following code in [urlMatcherFactory.js@0.2.15#L347](https://github.com/angular-ui/ui-router/blob/0.2.15/src/urlMatcherFactory.js#L347):
`result += encodeURIComponent(encoded);`
which takes the result of the custom type encoding function (`encoded`) and encodes it _again_ as a URIComponent, thereby switching the `/`s to `%2F`s in the resulting URL `result`.

So, as far I can tell, to fix this behaviour, we would need to find some way of contravening `encodeURIComponent`'s escaping of slashes, but only if the URL param's custom type says so. And how we would detect that, I don't know, and even if we did, it would be hacky at best. That's why I think we'd need some deeper, better solution here, or at least some kind of re-implementation of custom parameter typing.

**Comment by zivc** [comment]
Or we could implement a `ui-sref-unsafe`!

**Comment by mismith** [comment]
I guess the start to that would be delegating `encodeURIComponent` to the custom type's `encode` function itself rather applying it globally. Obviously this has security risks and would be a breaking change to existing custom types, but I would argue that custom types are an advanced feature and if a user was implementing them it would be easy to also enforce in the docs that it should return a URI-encoded result.

**Comment by perrierism** [comment]
@mismith thanks for figuring this out and pointing out exactly what's going on.

it seems like if types just had a flag to declare that they expect to do URI-encoding in the encode function:

```javascript
  $urlMatcherFactory.type("MyType", {
      performsAllNeededUriEncoding : true, // defaults to false
      encode: function(val) {return val != null ? val.toString() : val;  }
      decode: function(val) {return val != null ? val.toString() : val;  },
      is: function(val) {return this.pattern.test(val); },
      pattern: /[^/]+\/[^/]+/
    });
```
then urlMatcherFactory could just have a different treatment for this, and we wouldn't break existing custom types or need to change any of the built-in types:

```javascript
      ....
      if (squash === false) {
        if (encoded != null) {
          if (isArray(encoded)) {
            result += map(encoded, encodeDashes).join("-");
          } else {
            if (param.type.performsAllNeededUriEncoding){
                 result += ecoded;
            } else {
                 result += encodeURIComponent(encoded);
            }
          }
        }
      ...
```

maybe something similar is needed on the decode side

**Comment by mismith** [comment]
Great idea @perrierism ! It might be a bit hacky but at least it's fixing the issue. @nateabele if I threw a pull request together for this, what do you think the chances of it being merged are?

**Comment by christopherthielen** [comment]
Doubtful for 0.2.x due to BC, but I'd definitely consider changing the
contract for 1.0

**Comment by mismith** [comment]
Thanks @christopherthielen. Does the proposed change break backwards compatibility though?

**Comment by nateabele** [comment]
@mismith [Here you go](https://github.com/angular-ui/ui-router/pull/2218). I implemented it with a sane level of control structure nesting, and a sane key name. Open a PR against [feature-1.0](https://github.com/angular-ui/ui-router/tree/feature-1.0) with a failing test that passes on #2218 and I'll merge both.

**Comment by mismith** [comment]
@nateabele sweet, nicely done. I'm not sure I understand your request however; would you like me to write a test for #2218? I'm not using 1.0 nor am I familiar with TypeScript yet, so I don't know possible that is. Or did you mean I should write an equivalent feature for 0.2.x (including tests)? That I could do!

**Comment by nateabele** [comment]
> [W]ould you like me to write a test for #2218?

Yes.

> I'm not using 1.0 nor am I familiar with TypeScript yet, so I don't know possible that is.

The test suite & build setup are identical. You don't need to know anything about TypeScript. Just follow the contributor guidelines and add a failing test for a type with `raw: true`. Then run `grunt integrate`.

Let me know if you need help or have other questions.

**Comment by sanyam-j** [comment]
It this issue is fixed in latest release. If yes, then how i have to use it in defining ui routes. Bz i am getting the same issue with angular 1.4.7 & ui-router 0.2.15

**Comment by drothlis** [comment]
> If you really don't want the slash encoded for you, register a custom type with your regexp and declare item_id to be your custom type, i.e., url: /{item_id:MyType}

Note that since 0.2.15 ui-router provides a type called `any` that doesn't encode slashes (see 3bfd75a), so you can say `url: /{item_id:any}`.

**Comment by mismith** [comment]
@drothlis Unfortunately using 0.2.15 and `any` still doesn't fix the issue, at least not on Chrome (see my comments above https://github.com/angular-ui/ui-router/issues/1119#issuecomment-134686629 )

**Comment by drothlis** [comment]
@mismith I see, there are several subtly different issues here.

1. Some people want a *path* component in the URL to contain slashes, and for those slashes to show up in the URL as `/` not `%2F` (and certainly not `%252F`).
2. Some people want a *querystring* component in the URL to contain slashes. Are unescaped slashes are allowed here? Would you want them to show up in the URL as `/` or `%2F`? I'm not sure. For now, in my app, I've made my peace with them showing up as `%2F`.
3. Everybody wants either of the above to survive an encode/decode round trip. That is, if you generate a URL (e.g. by using `ui-sref('some-state', {some_param: "with/slashes"}`), then when you follow that link you should get the same parameter back from `$stateParams` (`"with/slashes"`, not `"with%2Fslashes"`).

(3) was broken because slashes were double-encoded in one direction and single-decoded in the other direction. Using the custom type (or `any` since 0.2.15 or 1.0.0alpha0) fixes this, at least in my use case: setting the parameter with `ui-sref` and getting it back from `$stateParams`. I don't use `$state.href()`.

(1) Still happens according to your [plunkr](http://plnkr.co/edit/CrD72FNYzLV1dATCjDdL?p=preview). In my use case (`any` + `ui-sref` + `$stateParams`) the URL shows correctly (with `/` instead of `%2F`) in the browser URL bar in Firefox and Chrome, but it still shows `%2F` in the status bar in Chrome (when you hover over the link). I assume that this is because the browser automatically converts `%2F` to `/` (but only in path components, not in querystrings). (?)

Honestly I can't tell which of the various github issues is supposed to track which of the above problems, but at least for my use case, I can confirm that `any` on 0.2.15 and 1.0.0alpha0 fixes (3) for me.

Thanks for the info, the plunkr, and the link to the new `raw` flag -- it all helped to clarify the above for me.

**Comment by sanyam-j** [comment]
@drothlis I am using "url: '/search/*path'". How can i use "any" with dynamic path?

**Comment by drothlis** [comment]
I should add that somewhere between 0.2.13 and 0.2.15, ui-router broke querystring components with slashes. In 0.2.13 you could have a state like this:

    url: "path?param1&param2"

In 0.2.15 that will no longer work if the parameter values have slashes in them: the encode/decode round trip will give back a different result than the original value, just like it was already broken for *path* components in 0.2.13.

In 0.2.15 you will have to use the `any` type like this:

    url: "path?{param1:any}&{param2:any}"

**Comment by christopherthielen** [comment]
Thanks for your patience.  Please build off `master` and see if this commit 02e9866 has addressed the issue for your apps and use cases.

**Comment by moneytree-doug** [comment]
Okay, so my particular case is that I am passing a redirect_url for OAuth. This is how it encodes:
```
https:%252F%252Fstage.example.com%252FHA%252Fapi%252Fcore.php
```
My current work around is to decode it twice before I do any further processing with it. Here it is after first decode: 

```
https:%2F%2Fstage.example.com%2FHA%2Fapi%2Fcore.php
```
After the second decode:
```
https://stage.example.com/HA/api/core.php
```

**Comment by christopherthielen** [comment]
@moneytree-doug urlencoding of the slashes was the root of the problem, which caused edge cases.  I switched away from urlencoding slashes to a custom encoding that the browser cannot pre-process without our knowledge.

Since I am not familiar with the OAuth flow, can you explain how your url parameter works through your workflow, and how the fix from 02e9866 is not helping?  The goal for ui-router is to allow parameter values including path parameters to have encoded slashes, to allow path parameters to be in the middle of the path, and to be encoded/decoded bi-directionally (allowing a url with slashes in parameters to be bookmarked)

As far as I can tell, we now perform this encoding properly. Slashes should encode to `~2F`, and decode back to `/`.    Tildes `~` should encode to `~~` and decode back to `~`.

**Comment by lpsBetty** [comment]
@christopherthielen I just upgraded from 0.2.15 to 0.2.18 and got this problem when using OAuth.. the slashes in the `redirectUri` parameter should be `%252F` instead of `~2F`

0.2.15 - works! 
````
http://localhost/accounts/test/projects/example-project/connect/github?redirectUri=%252Ftest%252Fexample-project%252Fsetup%252Fgithub
````

0.2.18
````
http://localhost/accounts/test/projects/example-project/connect/github?redirectUri=~2Ftest~2Fexample-project~2Fsetup~2Fgithub
````

this PR https://github.com/angular-ui/ui-router/pull/2071 is already in 0.2.18 I think?

**Comment by moneytree-doug** [comment]
@lpsBetty I had a similar situation. You should expect the tilde after upgrading to 0.2.18. I ran into tons of headaches with this issue in the versions before and I had monkey patches, so I had tests around the encoding part until the fix came out. I am also doing OAuth, and I have migrated it safely.

Your URL in the browser will definitely have a tilde, and this is fine and is expected. However, when you access it in your application code, ui-router will properly decode it in the `$stateParams` or `$state.params`. 

Correct me if I misunderstood.

**Comment by Hamcker** [comment]
thanks all about the comments and patience. I followed this issue to find a way to **avoid encoding url** buy what I got was **avoid to encode slashes**!!

my problem is the non-Latin chars. they're still being encoded. for example Arabic, Hebrew, Farsi etc.

for example consider I have a HTML like this :

    <a ui-sref="someState(Param:'آپارتمانی')"> A localized param </a>

which will be compiled to: 

    <a href="#!/Procuts/%D8%B2%DB%8C%D8%B1%20%D8%B2%D8%A7%D9%86%D9%88"> A localized param </a>

on StackOverflow:

- [StackOverflow Question](http://stackoverflow.com/questions/36427165/how-to-avoid-url-encoding-with-ui-sref)
- [Another StackOverflow Question](http://stackoverflow.com/questions/32867026/how-to-stop-ui-sref-encoding-url-params)

is it possible to avoid that?

**Comment by dkrugman** [comment]
tl;dr 

... but I see that there is still some activity here so I'm adding that upgrading from 0.2.10 to anything above 0.2.11 breaks my code, which depends on slashes in the URL.

After upgrading to 0.2.18:
the URL:     /browse/Children/Girls/Dresses+&+Skirts
becomes:    /browse/Children~2FGirls~2FDresses+&+Skirts

and I see that in the browser address bar:
![screenshot 2016-05-05 18 45 10](https://cloud.githubusercontent.com/assets/4335449/15059627/d6f9601a-12f1-11e6-98be-b5fafd59eabb.png)


So I'm sticking with 0.2.11, but my 2 cents is that whatever happened here over the last 2 years created a significant issue in that I now see the URL get encoded in 0.2.18.

I hope that's helpful in ultimately resolving this issue.

Thanks

**Comment by moneytree-doug** [comment]
@dkrugman Why is there no `#` in your URL?

**Comment by perrierism** [comment]
@moneytree-doug he has Html5Mode equal to True https://docs.angularjs.org/api/ng/provider/$locationProvider

**Comment by majew7** [comment]
My issue may have been different, but I was happy to find `:any` in this discussion.  I my code replaced 
`url: 'products/{slugPath:.*}'`
with
`url: 'products/{slugPath:any}'`

and I had success with finally seeing the forward slash when writing the slugPath rather than `~2F`.

**Comment by christopherthielen** [comment]
Also note we support non-encoded param types in 1.0, via `raw: true`

[Here's the docs](https://ui-router.github.io/docs/latest/classes/params.paramtype.html#raw) ;)

**Comment by tuoxiansp** [comment]
@majew7 Thanks!

**Comment by olehpanin** [comment]
@fabiosussetto BIG THANKS MAN!!!!!!

**Comment by tanyaburlakova** [comment]
hey @fabiosussetto!

I have another problem when I need to use param as query param. 
I can use it like url: '/login?next={nextUrl:nonURIEncoded}',
and I see this login?nextUrl=%2Fisv%2Fedit%2Ftanya-123%2Fpreview

Can u please help to understand what I need to do to fix it?
Or maybe some another way to resolve my task. I need to set param with previous url to link.
Thank u!
